google/cloud/mocks/current_options.h

```diff
@@ -0,0 +1,51 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_MOCKS_CURRENT_OPTIONS_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_MOCKS_CURRENT_OPTIONS_H
+
+#include "google/cloud/options.h"
+
+namespace google {
+namespace cloud {
+namespace mocks {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+/**
+ * Retrieve options used in a client call.
+ *
+ * This would be used to verify configuration options from within a
+ * MockConnection. It provides a way for applications to test the difference
+ * between `client.Foo(request, options)` and `client.Foo(request)`.
+ *
+ * @code
+ * TEST(Foo, CallOptions) {
+ *   auto mock = std::make_shared<MockConnection>();
+ *   EXPECT_CALL(*mock, Foo).WillOnce([] {
+ *         auto const& options = google::cloud::mocks::CurrentOptions();
+ *         EXPECT_THAT(options, ...);
+ *       });
+ *   auto client = Client(mock);
+ *   DoThing(client);
```

Comment:
Opt: Perhaps `client->Foo()` would be clearer?

Reply:
It's not quite the same thing. If a customer's test called Foo, they would just verify the options outside of the mock.

```cc
client.Foo(options);
EXPECT_THAT(options, ...);
```

We are helping them unit test their `DoThing` method which could look like:

```cc
void DoThing(Client c) {
  auto options = ...;
  c.Foo(options); // `c.Foo();` would be wrong, but how can I detect the difference?
}
```

That being said, I agree that the link between `DoThing` and `Foo` is unclear / imaginary. I will `s/DoThing/CallFoo/` to make the link clearer.

Reply:
decided on: `s/DoThing/MyFunctionThatCallsFoo/`

Reply:
SGTM.  Thanks.